internal/dynamic6provider/dynamic_element_type_test.go

```diff
@@ -0,0 +1,253 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package dynamic6provider_test
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	r "github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/knownvalue"
+	"github.com/hashicorp/terraform-plugin-testing/statecheck"
+	"github.com/hashicorp/terraform-plugin-testing/tfjsonpath"
+	"github.com/hashicorp/terraform-provider-corner/internal/testing/testprovider"
+	"github.com/hashicorp/terraform-provider-corner/internal/testing/testsdk/providerserver"
+	"github.com/hashicorp/terraform-provider-corner/internal/testing/testsdk/resource"
+)
+
+func Test_Dynamic_Attribute_ListType(t *testing.T) {
+	r.UnitTest(t, r.TestCase{
+		Steps: []r.TestStep{
+			{
+				// This may eventually be considered an invalid schema. It currently behaves as expected but may not be exposed to provider developers
+				// to avoid confusion. While the element type is dynamic, all elements must still have the exact same type, which Terraform will achieve
+				// by either performing a conversion (like below, converting 12345 to "12345"), or throw an error if conversion is impossible, for example:
+				//
+				// 		Error: Incorrect attribute value type
+				//
+				// 		  on terraform_plugin_test.tf line 12, in resource "corner_dynamic_thing" "foo":
+				// 		  12: 					attribute_with_dpt = ["hey", { number = 12345 }]
+				//
+				// 		Inappropriate value for attribute "attribute_with_dpt": all list elements must have the same type.
+				//
+				// Related issue: https://github.com/hashicorp/terraform/issues/34574
+				Config: `resource "corner_dynamic_thing" "foo" {
+					attribute_with_dpt = ["hey", 12345]
+				}`,
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownValue(
+						"corner_dynamic_thing.foo",
+						tfjsonpath.New("attribute_with_dpt"),
+						knownvalue.ListExact([]knownvalue.Check{
+							knownvalue.StringExact("hey"),
```

Comment:
I'm not sure it's documented in the context you're describing, but it is an interesting thought I also share.

Just observing the behavior of Terraform/cty it looks potentially arbitrary, but digging more into `cty` documentation, I believe it may just be following the only safe conversion route possible: https://github.com/zclconf/go-cty/blob/main/docs/convert.md#conversion-charts

- The literal representation of this specific test config is known to Terraform as: `tuple[string, number]`
- The schema we defined tells Terraform/cty that it needs to be converted to a `list`, which has a single element type
- Terraform needs to determine the safe route to converge `[string, number]` into a single element type. Which the only safe option is `string`

If we switch the config around to:
```terraform
resource "corner_dynamic_thing" "foo" {
  attribute_with_dpt = [12345, "hey"]
}
```

We receive the same element type as `string` from the msgpack config:
```json
{
  "attribute_with_dpt": [
    [
      "\"string\"",
      "12345"
    ],
    [
      "\"string\"",
      "hey"
    ]
  ]
}
```

It would be nice to get some confirmation on this behavior, so I will note this down as something to ask the core team so we can understand a little deeper 👍🏻